pkg/testutil/test_suite_runner.go

```diff
@@ -96,31 +95,55 @@ func (runner *TestSuiteRunner) runTestScenario(scenarioName string, fixtureCxt *
 		default:
 			panic(errors.New(fmt.Sprintf("unit under test: %s not supported", unitUnderTest)))
 		}
-		runner.assertExpectations(assert, expectationCxt, actual, err)
+		runner.assertExpectations(assert, expectation, actual, err)
 	})
 }
 
-//assertExpectations validates the actual outcome against expected outcome
-func (runner *TestSuiteRunner) assertExpectations(assert *assert.Assertions, expectationCxt *expectContext, actual acktypes.AWSResource, err error) {
-	if expectationCxt.err != nil {
-		assert.NotNil(err)
-		assert.Nil(actual)
-		assert.Equal(expectationCxt.err.Error(), err.Error())
-	} else if expectationCxt.latest == nil { // successful delete scenario
-		assert.Nil(err)
+/* assertExpectations validates the actual outcome against the expected outcome.
+	There are two components to the expected outcome, corresponding to the return values of the resource manager's CRUD operation:
+		1) the actual return value of type AWSResource ("expect.latest_state" in test_suite.yaml)
+		2) the error ("expect.error" in test_suite.yaml)
+	With each of these components, there are three possibilities in test_suite.yaml, which are interpreted as follows:
+		1) the key does not exist, or was provided with no value: no explicit expectations, don't assert anything
+		2) the key was provided with value "nil": explicit expectation; assert that the error or return value is nil
+		3) the key was provided with value other than "nil": explicit expectation; assert that the value matches the
+			expected value
+	However, if neither expect.latest_state nor error are provided, assertExpectations will fail the test case.
+ */
+func (runner *TestSuiteRunner) assertExpectations(assert *assert.Assertions, expectation *Expect, actual acktypes.AWSResource, err error) {
+	if expectation.LatestState == "" && expectation.Error == "" {
+		fmt.Println("Invalid test case: no expectation given for either latest_state or error")
+		assert.True(false)
+		return
+	}
+
+	// expectation exists for at least one of LatestState and Error; assert results independently
+	if expectation.LatestState == "nil" {
```

Comment:
Instead of checking for a string value called "nil", you could use the `omitempty` Go tag on the [`Expect.LatestState` and `Expect.Error` fields](https://github.com/aws-controllers-k8s/elasticache-controller/blob/f07e1b2a5184c65a1c0039fa20778f6436489757/pkg/testutil/test_suite_config.go#L60-L61) after making those fields' Go types a string pointer:

```go
// Expect contains the expected resource state or error after
// executing the operation under test. Either LatestState or Error
// should be non-nil.
type Expect struct {
	LatestState *string `json:"latest_state,omitempty"`
	Error       *string `json:"error,omitempty"`
}
```

And then here in the `assertExpectations()` you can simply check to see `if expectation.LatestStatus == nil`. That way, your `test_suite.yaml` is shorter and easier to read, since you only need to set either a `latest_state` or an `error` field, depending on whether the expected outcome is an error or not.

Reply:
Thanks for the suggestion re: the `omitempty` tag – my one concern is that, in addition to asserting LatestState and Error independently, we need to support the following behaviors:

1. Assert that LatestState or Error matches the given value
2. Assert that LatestState or Error are nil
3. Don't care about the value of LatestState or Error

The third case is the problematic one. There are some scenarios where the call to a resource manager's CRUD operation results in an error, but the returned resource is actually non-`nil` (e.g. the first test case `ReadOne=DNE` in `test_suite.yaml`). I think in this case we only care about asserting the error, and that we don't necessarily care about the `nil`-ness of the resource. 

Our initial implementation failed here because it assumed that every time the error was non-nil, the returned resource had to be `nil` (i.e both these assertions were coupled together in the same place). 

The block of code you suggested the below for asserting the `expectation.Error` should work with the `omitempty` approach, since we pretty much always care about the value of the Error, but how can we specify that we don't care about the value of `expectation.LatestState`? Changing the expectation struct to pointers to strings would allow us to do the following:

1. Omit the key entirely in test config -> `Expect.LatestState` is `nil`
2. Include the key with an empty value in test config -> `Expect.LatestState` is non-`nil`, `*Expect.LatestState` is `""`
3. Include the key with a nonempty value in test config -> `Expect.LatestState` is non-`nil`, *Expect.LatestState` is non-empty

We could map (1) to "explicitly assert nil" and (2) to "don't care about the value" to get rid of all the extra `nil` assertion lines. However, this would make the test config very confusing to read, and I really don't think we should sacrifice clarity for brevity in this case. 

Alternatively, we could use the more intuitive mapping of (1) to "don't care about the value" and (2) to "explicitly assert nil", but this would be just as long as the current implementation, and more confusing.

So in summary I think we should keep it as-is since it's the clearest/most intuitive way I can think of to encode the expectations of the test writer, and that the marginal gain of each scenario being 1-2 lines shorter in the test config is not worth the added confusion. Please let me know what you think or if I misinterpreted the suggestion!